diff --git a/gost_ameth.c b/gost_ameth.c
index 199aef2..9aea67c 100644
--- a/gost_ameth.c
+++ b/gost_ameth.c
@@ -121,6 +121,7 @@ static int gost_decode_nid_params(EVP_PKEY *pkey, int pkey_nid, int param_nid)
     case NID_id_GostR3410_2012_256:
     case NID_id_GostR3410_2012_512:
     case NID_id_GostR3410_2001:
+    case NID_id_GostR3410_2001DH:
         if (!key_ptr) {
             key_ptr = EC_KEY_new();
             if (!EVP_PKEY_assign(pkey, pkey_nid, key_ptr)) {
@@ -906,6 +907,7 @@ int register_ameth_gost(int nid, EVP_PKEY_ASN1_METHOD **ameth,
         return 0;
     switch (nid) {
     case NID_id_GostR3410_2001:
+    case NID_id_GostR3410_2001DH:
         EVP_PKEY_asn1_set_free(*ameth, pkey_free_gost_ec);
         EVP_PKEY_asn1_set_private(*ameth,
                                   priv_decode_gost, priv_encode_gost,
diff --git a/gost_eng.c b/gost_eng.c
index 69f9975..2a0ae70 100644
--- a/gost_eng.c
+++ b/gost_eng.c
@@ -88,6 +88,7 @@ static int gost_digest_nids(const int** nids) {
 
 static int gost_pkey_meth_nids[] = {
         NID_id_GostR3410_2001,
+        NID_id_GostR3410_2001DH,
         NID_id_Gost28147_89_MAC,
         NID_id_GostR3410_2012_256,
         NID_id_GostR3410_2012_512,
@@ -390,6 +391,7 @@ static int gost_pkey_meths(ENGINE* e, EVP_PKEY_METHOD** pmeth,
 
     switch (nid) {
         case NID_id_GostR3410_2001:
+        case NID_id_GostR3410_2001DH:
             *pmeth = pmeth_GostR3410_2001;
             return 1;
         case NID_id_GostR3410_2012_256:
@@ -433,6 +435,7 @@ static int gost_pkey_asn1_meths(ENGINE* e, EVP_PKEY_ASN1_METHOD** ameth,
 
     switch (nid) {
         case NID_id_GostR3410_2001:
+        case NID_id_GostR3410_2001DH:
             *ameth = ameth_GostR3410_2001;
             return 1;
         case NID_id_GostR3410_2012_256:
